### ISSUE TITLE
Global Pointer value analysis is not run after rebasing (originally: MIPS rebasing doesn't adjust $gp)

### ISSUE THREAD

**Issue by lwerdna** [Vector35/binaryninja-api#4556]
**Version and Platform (required):**
3.5

**Bug Description:**
MIPS rebasing doesn't adjust $gp, references to $gp-sensitive stuff is lost on rebase, like the global offset table.

**Steps To Reproduce:**
Open the attached binary: [libhdk-mips-elf.zip](https://github.com/Vector35/binaryninja-api/files/12284520/libhdk-mips-elf.zip)

Navigate to the memcpy plt stub:

```
00018900  int32_t memcpy()
00018900  1080998f   lw      $t9, -0x7ff0($gp)  {0x0}  {_GLOBAL_OFFSET_TABLE_}
```

It correctly gets the pointer to the \_GLOBAL\_OFFSET\_TABLE\_ at 0x2c400 using $gp and we can check the math in the python console:

```
>>> hex(bv.global_pointer_value.value - 0x7ff0)
'0x2c400'
>>> hex(bv.symbols['_GLOBAL_OFFSET_TABLE_'][0].address)
'0x2c400'
```

Now rebase the program to any address, say 0x1000 and the reference to the GOT is lost:

```
00019900  int32_t memcpy()
00019900  1080998f   lw      $t9, -0x7ff0($gp)  {data_2c400}
```

**Expected Behavior:**

The reference to \_GLOBAL\_OFFSET\_TABLE\_ is maintained after rebasing.

**Comment by plafosse** [comment]
Current workaround for this is to open the binary with options rather than rebasing after open.

**Comment by xusheng6** [comment]
I tested and found that with my recent changes on the global pointer value update, the global pointer value is already corrected updated after the rebasing. However, the function is not re-analyzed so it still has the old analysis. If I manually update the analysis then it works. The next step is to understand why we missed an analysis update for it

**Comment by xusheng6** [comment]
We are missing the analysis update because once we decide a function uses the gp value and update its analysis accordingly, the function's argument list no longer contains the gp register, and it is no longer considered using the gp. In other words, our "function uses gp" detection is only one-shot, when the gp value changes, either because the entire analysis state change, or, in the future, when a gp value is specified by the user (#5648), these functions all would not be updated properly.

**Comment by xusheng6** [comment]
There is a regression on this before 4.2.6430-dev (d7f78e8c)

**Comment by xusheng6** [comment]
This is more complex than I have expected. Here is a brief summary of what is happening:

1. When the rebase happens, we created a new binary view (and a new analysis object), so the GP value update analysis is actually properly run
2. However, due to another bug (see below) related to relocation handling, one of the functions are disassembled erroneously. And this errors happens to affect the GP value calculation of that function. As a result, the new GP value, despite being correctly calculated, lose to the old since it has fewer voters. 
3. This is already wrong on the last stable (for the same reason). I remember it once worked (at least when I checked in the commit that fixes this), but it is broken soon after. We really need to add a unit test for this
4. For the actual bug, look at the code at 0x5660 before rebasing:

![Screenshot 2024-11-14 at 5 15 00 PM](https://github.com/user-attachments/assets/501e094c-d3b0-4782-ab48-c82d3d5d0916)

we can see it calls `sub_15e60`. However, after rebasing the binary to 0x10000, it becomes this:

![Screenshot 2024-11-14 at 5 15 54 PM](https://github.com/user-attachments/assets/8cda36bf-a316-4344-9443-7d9279f253c7)

We can see it now references the data variable at `data_3c43c` correctly, instead of `data_2c43c`, but the value at data_2c43c still points to `0x15e60`, not `0x25e60`

![Screenshot 2024-11-14 at 5 16 47 PM](https://github.com/user-attachments/assets/a90f893e-d077-4896-9575-d2adb314c76a)
